doc/design/audit-trail-mvp.md

```diff
@@ -0,0 +1,152 @@
+example, when a user publishes a Dandiset, the published `Version` that results
+can be understood as originating from the draft version present in the Dandiset
+at a specific time. Taken in that sense, these simple provenance events can be
+identified with corresponding audit events.
+
+More complex provenance events do not currently exist in the system. For
+example, a reasonable action a user could take would be to manufacture a new
+Dandiset from the state of a published version in some other Dandiset. If this
+action were afforded to the user, we might think of recording a provenance event
+describing the specific way in which that derived Dandiset came to be.
+
+Because we do not currently afford such complex provenance events in the system,
+the operations to fulfill the basic audit system will be sufficient to support
+any provenance events in the future. See the [Operations](#operations) section below for
+more details.
+
+## Operations
+
+In addition to providing audit-style trails for Dandisets and assets, this plan
+fulfills a secondary requirement to cut the `previous` links, freeing up
+non-current assets for garbage collection. As this will destroy the asset chains
+from which we might draw some limited historical information, we need to
+carefully design an operations plan that preserves maximal information while
+empowering the garbage collection work to continue.
+
+Because it is broadly impossible to glean all of the information needed to
+reconstruct the historical audit trail for existing and deleted Dandisets, we
+plan to *begin* collecting complete information with the implementation
+of this design, while **preserving** a database snapshot at that
+same moment so that, if we choose in the future, we may wish to reconstruct that
+partial historical record. We would be essentially splitting time into two
+periods: before-audit-trail (represented by the DB snapshot) and
+after-audit-trail (represented by live, ongoing collection of complete audit
+records).
+
+This plan does not negatively impact future efforts at building
+provenance-related features. If we need to generalize provenance events from the
+audit trail, we can do so from the after-audit-trail events as needed, and the
+before-audit-trail DB snapshot remains available for that purpose as well. In
+essence, our plan retains the maximum possible amount of information while
+freeing the system to perform new features like audit, provenance, and
+comprehensive garbage collection.
+
+In sum, this plan allows us to (1) implement audit into perpetuity sooner; (2)
+preserve the available record at the moment that goes into effect; and (3)
+enable the completion of garbage collection design and implementation.
+
+The sequence of events looks like this:
+
+1. Implement the design in this document as a pull request, but do not
+   merge/release it yet (upon merge and release, the Archive will begin
+   generating audit records).
+2. Put the system into maintenance mode.
```

Comment:
Is there a doc for what mechanism will be used to implement this system? (e.g. is the audit log maintained by some Django super.save() call? Inspecting the incoming auth token to infer the user, etc. etc.?)

Reply:
There is domain knowledge about the codebase in terms of internal services we would use to identify, capture, and export this information to the database. It is too fine-grained to include in this doc, which largely is concerned with the concepts, requirements, and operations concerns involved.

If you are interested in these finer details, you can take a look at the eventual implementation, or else I would be happy to meet with you to discuss.

Reply:
I'll stay tuned for the implementation then -- thanks Roni!